wire/msgmixslotreserve.go

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) 2023 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package wire
+
+import (
+	"fmt"
+	"hash"
+	"io"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+const (
+	// MaxMixMcount is the maximum number of mixed messages that are allowed
+	// in a single mix.  This restricts the total allowed size of the slot
+	// reservation and XOR DC-net matrices.
+	MaxMixMcount = 1024 // XXX: PNOOMA
+
+	// MaxMixKPCount is the maximum number of peers allowed together in a
+	// single mix.  This restricts the total size of the slot reservation
+	// and XOR DC-net matrices.
+	MaxMixKPCount = 512 // XXX: PNOOMA
+
+	// MaxMixFieldValLen is the maximum number of bytes allowed to represent
+	// a value in the slot reservation mix bounded by the field prime.
+	MaxMixFieldValLen = 32
+)
+
+// MsgMixSlotReserve is the slot reservation broadcast.  It implements the Message
+// interface.
+type MsgMixSlotReserve struct {
+	Signature       [64]byte
+	Identity        [33]byte
+	SessionID       [32]byte
+	Expiry          uint32
+	Run             uint32
+	DCMix           [][][]byte // mcount-by-peers matrix of field numbers
+	SeenCiphertexts []chainhash.Hash
+
+	// hash records the hash of the message.  It is a member of the
+	// message for convenience and performance, but is never automatically
+	// set during creation or deserialization.
+	hash chainhash.Hash
+}
+
+// BtcDecode decodes r using the Decred protocol encoding into the receiver.
+// This is part of the Message interface implementation.
+func (msg *MsgMixSlotReserve) BtcDecode(r io.Reader, pver uint32) error {
+	const op = "MsgMixSlotReserve.BtcDecode"
+	if pver < MixVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	err := readElements(r, &msg.Signature, &msg.Identity, &msg.SessionID,
+		&msg.Expiry, &msg.Run)
+	if err != nil {
+		return err
+	}
+
+	// Read the DCMix
+	mcount, err := ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+	kpcount, err := ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+	if mcount == 0 {
```

Comment:
These mcount checks probably should come before reading the kpcount since there is no reason to read on a guaranteed failure.